crates/net/network/src/transactions.rs

```diff
@@ -1074,116 +1255,287 @@ struct Peer {
-    /// Set that tracks all hashes that are currently being fetched.
-    inflight_hash_to_fallback_peers: HashMap<TxHash, Vec<PeerId>>,
+    /// Hashes that are awaiting fetch from an idle peer.
+    buffered_hashes: LruCache<TxHash>,
+    /// Tracks all hashes that are currently being fetched or are buffered, mapping them to
+    /// request retries and last recently seen fallback peers (max one request try for any peer).
+    unknown_hashes: LruMap<TxHash, (u8, LruCache<PeerId>), Unlimited>,
+    /// Size metadata for unknown eth68 hashes.
+    eth68_meta: LruMap<TxHash, usize, Unlimited>,
 }
 
 // === impl TransactionFetcher ===
 
 impl TransactionFetcher {
     /// Removes the specified hashes from inflight tracking.
     #[inline]
-    fn remove_inflight_hashes<'a, I>(&mut self, hashes: I)
+    fn remove_from_unknown_hashes<I>(&mut self, hashes: I)
     where
-        I: IntoIterator<Item = &'a TxHash>,
+        I: IntoIterator<Item = TxHash>,
     {
-        for &hash in hashes {
-            self.inflight_hash_to_fallback_peers.remove(&hash);
+        for hash in hashes {
+            self.unknown_hashes.remove(&hash);
+            self.eth68_meta.remove(&hash);
         }
     }
 
-    /// Advances all inflight requests and returns the next event.
-    fn poll(&mut self, cx: &mut Context<'_>) -> Poll<FetchEvent> {
-        if let Poll::Ready(Some(GetPooledTxResponse { peer_id, requested_hashes, result })) =
-            self.inflight_requests.poll_next_unpin(cx)
-        {
-            return match result {
-                Ok(Ok(transactions)) => {
-                    // clear received hashes
-                    self.remove_inflight_hashes(transactions.hashes());
+    /// Updates peer's activity status upon a resolved [`GetPooledTxRequest`].
+    fn update_peer_activity(&mut self, resp: &GetPooledTxResponse) {
+        let GetPooledTxResponse { peer_id, .. } = resp;
 
-                    // TODO: re-request missing hashes, for now clear all of them
-                    self.remove_inflight_hashes(requested_hashes.iter());
+        debug_assert!(
+            self.active_peers.get(peer_id).is_some(),
+            "broken invariant `active_peers` and `inflight_requests`"
+        );
 
-                    Poll::Ready(FetchEvent::TransactionsFetched {
-                        peer_id,
-                        transactions: transactions.0,
-                    })
+        let mut remove = || -> bool {
+            if let Some(inflight_count) = self.active_peers.get(peer_id) {
+                if *inflight_count <= 1 {
+                    return true
                 }
-                Ok(Err(req_err)) => {
-                    // TODO: re-request missing hashes
-                    self.remove_inflight_hashes(&requested_hashes);
-                    Poll::Ready(FetchEvent::FetchError { peer_id, error: req_err })
+                *inflight_count -= 1;
+            }
+            false
+        };
```

Comment:
this closure is always evaluated, so we don't need to put this into a closure and can evaluate it directly?

Reply:
I just do it like this sometimes to save LOC and because it's nice to read, easier to see directly that one branch generates `true`, all others generate `false`
```rust
let remove = if let Some(inflight_count) = self.active_peers.get(peer_id) {
    if *inflight_count <= 1 {
        true
    } else {
        *inflight_count -= 1;
         false
    }
} else {
    false
};
```
```rust
let remove = || -> bool {
    if let Some(inflight_count) = self.active_peers.get(peer_id) {
        if *inflight_count <= 1 {
            return true
        }
        *inflight_count -= 1;
    }
    false
}();
```

Reply:
since it's not a move closure, it doesn't get its own stack frame, so it's not costly